### PR TITLE
[DRAFT, Enhancement] Skip egg hatching and show summary

### PR DESCRIPTION
**Pull request by j-diefenbach**
## What are the changes?

- Add a prompt that enables players to skip egg hatching and go straight to a summary page
- Adding an egg summary page that gives a quick overview of the pokemon hatched from eggs:
- Use icons to indicate if a hatched pokemon had a hidden ability, egg move or was shiny (alongside their actual sprite)
- Animations (jumping) for new pokemon not caught before
- Able to navigate pokemon with cursor to see pokemon sprite and info (same as when caught/hatched normally)

## Why am I doing these changes?

* Egg hatching can be tedious and time consuming, watching animations and dialogue for mostly unimportant pokemon
* The ease of skipping these animations altogether can save large amounts of time
* The summary page provides players the opportunity to completely skip egg hatching and still be able to check if there were any noticeable hatches / unlocks at a glance incase they missed something
* Improves overall QoL for those hatching large amounts of eggs

#### What changes for players?
- when eggs hatch a "skip egg hatching?" prompts the player for when there are more than 5 eggs
- there is a new egg summary screen which shows various icons for pokemon hatched, and can be navigated with a cursor to see more info for each player (exited with 'X' / cancel)
#### What changes for the code?
- Added an EggSkipPhase that determines whether there is enough eggs hatching to be skipped, parses the "skip egg hatching?" prompt to either serially start EggHatchPhases (as per main) or start the EggSummaryPhase
* Added an EggSummaryPhase which performs hatching and pokedex updates
* Added an EggSummaryUiHandler which manages the UI of all the icons, info screens, sprites, cursors etc. (the UI logic was moved from EggSummaryPhase to EggSummaryUiHandler) - allows cursor navigation and updates the current display info and sprite accordingly, can exit with Button.CANCEL

## TODOs (feedback is welcome, especially for UI and style/layout)
- [X] Add skip prompt
- [X] Seperate egg hatching functionality from animations
- [ ] Fix egg hatching and UI setup lag / improve load times or appropriate message / animation
- [X] Summary page broad layout
- [ ] Neat icons
- [ ] Clarify what icons mean what and appropriate logic:
  - [X] Hidden ability icon
  - [X] Shiny icon
  - [ ] New Egg Move icon
  - [ ] Pokeball / new unlock icon (needs to be clarified)
  - [ ] Catch animation (new pokemon jump)
- [X] Implement UI navigation for summary
- [X] Logic for storing new catch / hatch info and new egg moves
- [ ] Proper UI to show new unlocks / hatch info (as per showInfo function) on left side of screen
  - [ ] Display (and fit) pokemon sprite on left side of screen above info
  - [ ] New background might be necessary (currently reuses eggListBg, has weird shadow built in and may not fit sprites well)

### Screenshots/Videos

Large amount of hatched pokemon, pokemon sprite working and showInfo working (showInfo still needs to be reworked / text taken out and rearranged to fit - this is temporary to show catch info and prove info is working as intended)
image

Note there is a large lag when the pokemon are being hatched / generated and the UI is being setup. I believe it's due to the overhead of loading the showInfo screens beforehand so it isn't overwritten by new pokedex info.
EggHatchSummaryGif

## How to test the changes?

Git checkout the branch, overrides are currently on the branch. Pull some eggs (manual free egg override atm), then finish any wave and accept prompt)
- Currently the eggs do not clear since it is less of a hassle to setup to check minor changes

## Checklist
- [X] Have I checked that there is no overlap with another PR?
- [X] Have I made sure the PR is self-contained and cannot be split into smaller PRs?
- [X] Have I provided a clear explanation of the changes?
- [ ] Have I tested the changes (manually)? (NOT READY FOR TESTING)
  - Currently the eggs do not clear since it is less of a hassle to setup to check minor changes
- [X] Are the changes visual?
  - [X] Have I provided screenshots/videos of the changes?
  - Note visual changes WIP and layout, style not decided or close to finished

### CODE REVIEW

**Comment by TempsRay** [comment]
👀

**Comment by H-i-eu** [comment]
this would be so efficient in reducing all that time hatching i appreciate this so much HAHA

**Comment by Arxalc** [comment]
I was messing around with trying to skip hatching animations, but I don't have anything concrete yet. I'm commenting here with my thoughts in hopes they might help someone else looking into it.

`EggLapsePhase` in `phases.ts` "holds" all the individual egg hatching phases. 
`EggHatchPhase `in `egg-hatch-phase.ts` runs through one egg's hatching process.
If you want to skip animations and just update game data, I believe you pretty much only just need this part in `egg-hatch-phase.ts`
`this.scene.gameData.updateSpeciesDexIvs(this.pokemon.species.speciesId, this.pokemon.ivs);
          this.scene.gameData.setPokemonCaught(this.pokemon, true, true).then(() => {
            this.scene.gameData.setEggMoveUnlocked(this.pokemon.species, this.eggMoveIndex)
...`

**_How do you know if/when to skip animations? That's my problem atm._**
**Pressing 'Cancel' during the animation to skip all animations** - Button input processing during egg hatching is handled in `egg-hatch-scene-handler.ts` . Idk how to... propogate? an input from an individual phase to the EggLapsePhase to then apply to the rest of the phases. Not sure if I can return a value or make a public/static variable to look out for it.
**Have an option in Settings to skip hatching animation** - Might be easier to do. Get the settings option, pass as a constructor parameter into EggHatchPhase, and go from there?

**Comment by j-diefenbach** [comment]
> I was messing around with trying to skip hatching animations, but I don't have anything concrete yet. I'm commenting here with my thoughts in hopes they might help someone else looking into it.
> 
> `EggLapsePhase` in `phases.ts` "holds" all the individual egg hatching phases. `EggHatchPhase `in `egg-hatch-phase.ts` runs through one egg's hatching process. If you want to skip animations and just update game data, I believe you pretty much only just need this part in `egg-hatch-phase.ts` `this.scene.gameData.updateSpeciesDexIvs(this.pokemon.species.speciesId, this.pokemon.ivs); this.scene.gameData.setPokemonCaught(this.pokemon, true, true).then(() => { this.scene.gameData.setEggMoveUnlocked(this.pokemon.species, this.eggMoveIndex) ...`
> 
> **_How do you know if/when to skip animations? That's my problem atm._** **Pressing 'Cancel' during the animation to skip all animations** - Button input processing during egg hatching is handled in `egg-hatch-scene-handler.ts` . Idk how to... propogate? an input from an individual phase to the EggLapsePhase to then apply to the rest of the phases. Not sure if I can return a value or make a public/static variable to look out for it. **Have an option in Settings to skip hatching animation** - Might be easier to do. Get the settings option, pass as a constructor parameter into EggHatchPhase, and go from there?

My idea isn't about skipping individual egg hatches, instead you skip all the actual egg hatching phases so there is no animations. Instead the prompt will be inside eggLapsePhase before EggHatchPhase is called. This will then call some egg hatching functionality independent of animations (so that the pokemon and achievements are updated accordingly), then move to a summary page.
To clarify
1. Egg lapse phase message "Oh...?"
2. Prompt appears something like "25 eggs are hatching would you like to skip? y/n"
3. Yes -> skips egg hatchphase loop
4. Calls seperate functionality to hatch eggs without animations (will be identical to if there was animations)
5. Summary page is shown to display all the pokemon with relevant info / icons at once (see screenshots)
6. Possibly: can navigate summaries to see the showInfo screen with IVs abilities and more (i know there is some additions to it being planned)
7. Can press continue at any time in the summary page to exit

**Comment by flx-sta** [comment]
I'm just gonna link this here
https://github.com/pagefaultgames/pokerogue/issues/2202

**Comment by KAMJAIM** [comment]
Did this actually apply to the game?

**Comment by HannahShinatsuri** [comment]
Wanna ask cause am really newbie with all of this but how do u impor tthis to ur game? Honestly it would help and be big time gain..

**Comment by nayfaan** [comment]
are there any updates on this? is there a reason why this PR isn't merged yet?

**Comment by flx-sta** [comment]
@j-diefenbach are you still working on this?

**Comment by j-diefenbach** [comment]
> @j-diefenbach are you still working on this?

Yep, mostly just the UI and some tweaks to the way I show the info container so u can still see new unlocks even if the pokedex was updated

File: src/ui/pokemon-info-container.ts [hunk 1]
```diff
@@ -133,7 +134,7 @@ export default class PokemonInfoContainer extends Phaser.GameObjects.Container {
 
     this.statsContainer = new StatsContainer(this.scene, -48, -64, true);
 
-    this.add(infoBg);
+    this.add(this.infoBg);
```
Comment (frutescens) [inline src/ui/pokemon-info-container.ts:137]: Move this line to above line 133 and I think it should fix the problem.

**Review by frutescens — COMMENTED** [review]
Hope this fixes the problem.

**Comment by flx-sta** [comment]
Idk if merging `main` was a good idea

**Comment by j-diefenbach** [comment]
> Idk if merging `main` was a good idea. Normally you should only ever merge `beta`

yeah I forgot that my fork main was tracking main not beta ;-;

**Comment by flx-sta** [comment]
> > Idk if merging `main` was a good idea. Normally you should only ever merge `beta`
> 
> yeah I forgot that my fork main was tracking main not beta ;-;

As it should. You should have your own “beta” branch that tracks beta. Or simply always keep your upstream updated and merge from “upstream/beta”.


Otherwise it might look to other ppl like you are merging “main” even though your forks “main” is actually “beta”. Which is confusing

**Pull request closed without merging**